### DISCUSSION TITLE
TypeError: switchNetwork is not a function when using in useEffect

### DISCUSSION THREAD

**Discussion by 0xCardiE** [wevm/wagmi#1283] [Q&A] [answered]
I am trying something rather simple, I want that I use switchNetwork on page load if its on wrong network but I keep gettin error from subject. Why is this so I dont understand?

```
  useEffect(() => {
      if (chain.id !== process.env.NEXT_PUBLIC_NID) {
         switchNetwork(process.env.NEXT_PUBLIC_NID);
       }
  }, [])
```

**Comment by tmm** [comment]
Not all wallets support programmatic chain switching so you should check if `switchNetwork` is defined before using it (also drop the `await` too):

```ts
useEffect(() => {
  if (chain?.id !== process.env.NEXT_PUBLIC_NID) {
    switchNetwork?.(process.env.NEXT_PUBLIC_NID);
  }
}, [switchNetwork])
```

**Reply by 0xCardiE** [reply]
(Sorry **await** was typing mistake, code doesnt even run with await as it cant be used that way in useEffect.)

I am using just metamask for testing this which does support it right?

**Reply by 0xCardiE** [reply]
I realized this works if I dont refresh whole page, but if I change this code and hot reloading is done or whatever it is called when page is updated with changes then wallet switching is trigered, but if I make classic browser refresh and its all loaded then I keep getting this "switchNetwork" is not a function. Any ideas what could be the reason?

**Reply by tmm** [reply]
You need to add `switchNetwork` to the `useEffect` dependency array, the it should work.

**Reply by 0xCardiE** [reply]
Seems this also doesn't change much. Dont get it why is it saying this on page load

![image](https://user-images.githubusercontent.com/8969767/201434704-902fe10c-2041-4e50-b904-deee889a6df3.png)

**Reply by tmm** [reply]
It's possible that `switchNetwork` isn't defined on page load while connector is still (auto)connecting. You should check that it is defined before calling ([see previous comment](https://github.com/wagmi-dev/wagmi/discussions/1283#discussioncomment-4118787)).

**Reply by 0xCardiE** [reply]
Should I then try something like

```
  const connect = useConnect({
    onSuccess(data) {
      switchNetwork(process.env.NEXT_PUBLIC_NID);
    },
  });
```

so it fires on Connect success?  If so how do I fire this? I am struggling to understand how I could make this construct be triggered?

**Reply by 0xCardiE** [reply]
This works fine AFAIK

```
  useEffect(() => {
    if (chain?.id !== process.env.NEXT_PUBLIC_NID) {
      switchNetwork?.(process.env.NEXT_PUBLIC_NID);
    }
  }, [isConnected]);
```

**Comment by 0xCardiE** [comment, marked as answer]
This works fine AFAIK

```
  useEffect(() => {
    if (chain?.id !== process.env.NEXT_PUBLIC_NID) {
      switchNetwork?.(process.env.NEXT_PUBLIC_NID);
    }
  }, [isConnected]);
```
